src/filter_tree/filter_tree.c

```diff
@@ -439,89 +434,109 @@ static inline void _FilterTree_In_Place_Set_Exp(FT_FilterNode *node, SIValue v)
 static bool _FilterTree_Compact_And(FT_FilterNode *node) {
 	// Try to compact left and right children.
 	bool is_lhs_const = FilterTree_Compact(node->cond.left);
-	bool ihs_rhs_const = FilterTree_Compact(node->cond.right);
+	bool is_rhs_const = FilterTree_Compact(node->cond.right);
 	// If both are not compactable, this node is not compactable.
-	if(!is_lhs_const && !ihs_rhs_const) return false;
-	// Compact to 'false' in case at least one child is 'false'
-	if(is_lhs_const) {
-		SIValue lhs_value = AR_EXP_Evaluate(node->cond.left->exp.exp, NULL);
+	if(!is_lhs_const && !is_rhs_const) return false;
+	// In every case from now, there will be a reduction, save the children in local placeholders for current node in-place modifications.
+	FT_FilterNode *lhs = node->cond.left;
+	FT_FilterNode *rhs = node->cond.right ;
+	// Both children are constants. This node can be set as constant expression.
+	if(is_lhs_const && is_rhs_const) {
+		// Both children are now contant expressions. We can evaluate and compact.
+		SIValue rhs_value = AR_EXP_Evaluate(rhs->exp.exp, NULL);
+		SIValue lhs_value = AR_EXP_Evaluate(lhs->exp.exp, NULL);
+		SIValue final_value = SI_BoolVal(SIValue_IsTrue(lhs_value) && SIValue_IsTrue(rhs_value));
+		_FilterTree_In_Place_Set_Exp(node, final_value);
+		FilterTree_Free(lhs);
+		FilterTree_Free(rhs);
+		return true;
+	} else if(is_lhs_const) {
+		// Only lhs is constant - evaluate and free lhs.
+		SIValue lhs_value = AR_EXP_Evaluate(lhs->exp.exp, NULL);
+		FilterTree_Free(lhs);
 		// If lhs is false, everything is false.
-		if(!lhs_value.longval) {
-			FilterTree_Free(node->cond.left);
-			FilterTree_Free(node->cond.right);
+		if(SIValue_IsFalse(lhs_value)) {
 			_FilterTree_In_Place_Set_Exp(node, lhs_value);
-			return true;
-		}
-		// Lhs is true. Check Rhs.
-		if(ihs_rhs_const) {
-			// Rhs is const, there will be a compaction.
-			SIValue rhs_value = AR_EXP_Evaluate(node->cond.right->exp.exp, NULL);
-			FilterTree_Free(node->cond.left);
-			FilterTree_Free(node->cond.right);
-			// Since lhs is true, the new value is determent be rhs.
-			_FilterTree_In_Place_Set_Exp(node, rhs_value);
+			FilterTree_Free(rhs);
 			return true;
 		} else {
-			// Cannot compact;
+			// Lhs is true. Current node should be replaced with rhs.
+			memcpy(node, rhs, sizeof(FT_FilterNode));
```

Comment:
Shouldn't this: `memcpy(node, rhs, sizeof(FT_FilterNode));` be the implementation of `_FilterTree_In_Place_Set_Exp` ?

Reply:
No
the `memcpy` can copy a predicate or cond node data completely while `_FilterTree_In_Place_Set_Exp` actually transforms a cond or pred node into an expression.